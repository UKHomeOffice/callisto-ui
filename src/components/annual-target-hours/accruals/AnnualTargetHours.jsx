import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';

const AnnualTargetHours = ({ targetData, accrualsData }) => {
  const { t } = useTranslation('common');
  let total = targetData?.targetTotal || 0;
  let worked = accrualsData?.cumulativeTotal || 0;
  let remainingMins = Math.floor(total - worked);
  let remainingWhole = Math.floor((remainingMins) / 60);
  let target = total - (accrualsData?.cumulativeTarget || 0);

  if (!targetData) {
    total = '-';
    worked = '-';
    remainingMins = '-';
    target = '-';
  } else {
    total = formatToStringTime(total);
    worked = formatToStringTime(worked);
    remainingMins = formatToStringTime(remainingMins);
    target = formatToStringTime(target);
  }

  return (
    <div className="accruals-container">
      <form className="grey-border">
        <h1 className="govuk-heading-m newline">
          {targetData ? t('annualTargetHours.titleRemaining', { count: remainingWhole }) : t('annualTargetHours.noAgreement')}
        </h1>

        <table className="govuk-table">
          <tbody className="govuk-table__body">
            <tr className="govuk-table__row">
              <th scope="row" className="govuk-table__header">
                {t('annualTargetHours.total')}
              </th>
              <td className="govuk-table__cell govuk-table__cell--numeric">
                {total}
              </td>
            </tr>
            <tr className="govuk-table__row">
              <th scope="row" className="govuk-table__header">
                {t('annualTargetHours.worked')}
              </th>
              <td className="govuk-table__cell govuk-table__cell--numeric">
                {worked}
              </td>
            </tr>
            <tr className="govuk-table__row">
              <th scope="row" className="govuk-table__header">
                {t('annualTargetHours.remaining')}
              </th>
              <td className="govuk-table__cell govuk-table__cell--numeric">
                {remainingMins}
              </td>
            </tr>
            <tr className="govuk-table__row">
              <th scope="row" className="govuk-table__header">
                {t('annualTargetHours.target')}
              </th>
              <td className="govuk-table__cell govuk-table__cell--numeric">
                {target}
              </td>
            </tr>
          </tbody>
        </table>
      </form>
    </div>
  );
};

const formatToStringTime = (decimalHours) => {
  if (decimalHours > 0) {
    const hours = Math.trunc(decimalHours / 60);
    let minutes = decimalHours % 60;
    minutes = minutes.toString().padStart(2, '0');
    return `${hours}:${minutes}`;
  }
  return '';
};

export default AnnualTargetHours;

AnnualTargetHours.propTypes = {
  targetData: PropTypes.shape({
    id: PropTypes.string,
    tenantId: PropTypes.string,
    agreementId: PropTypes.string,
    accrualTypeId: PropTypes.string,
    targetTotal: PropTypes.number,
  }),
  accrualsData: PropTypes.shape({
    id: PropTypes.string,
    tenantId: PropTypes.string,
    personId: PropTypes.string,
    agreementId: PropTypes.string,
    accrualDate: PropTypes.string,
    accrualTypeId: PropTypes.string,
    cumulativeTotal: PropTypes.number,
    cumulativeTarget: PropTypes.number,
    contributions: PropTypes.shape({
      timeEntries: PropTypes.object,
      total: PropTypes.number,
    }),
  }),
};
