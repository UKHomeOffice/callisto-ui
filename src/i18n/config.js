import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import LanguageDetector from 'i18next-browser-languagedetector';
import Backend from 'i18next-http-backend';
import {
  format as formatDate,
  formatRelative,
  formatDistance,
  isDate,
} from 'date-fns';
import { enGB, fr } from 'date-fns/locale';
//import { DateTime } from 'luxon';

const locales = { enGB, fr };
i18n
  .use(Backend)
  .use(LanguageDetector)
  .use(initReactI18next)

  .init({
    debug: true,
    fallbackLng: 'en',
    interpolation: {
      escapeValue: false,
      format: (value, format, lng) => {
        if (isDate(value)) {
          const locale = locales[lng];
          //return formatDate(value, format, { locale });
          return formatDate(value, 'PPPP', { locale });
        }
      },
    },
    backend: {
      loadPath: '/locales/{{lng}}/{{ns}}.json',
    },
  });

//i18n.languages = ['en', 'fr'];
// i18n.services.formatter.add('DATE_HUGE', (value, lng, options) => {
//   return DateTime.fromJSDate(value)
//     .setLocale(lng)
//     .toLocaleString(DateTime.DATE_HUGE);
// });
export default i18n;
