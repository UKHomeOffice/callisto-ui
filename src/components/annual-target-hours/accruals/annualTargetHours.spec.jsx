import { render, screen } from '@testing-library/react';
import AnnualTargetHours from './AnnualTargetHours';
import { initReactI18next } from 'react-i18next';
import i18n from 'i18next';
import { withTranslation, useTranslation } from 'react-i18next';
import { I18nextProvider } from 'react-i18next';
import '../../../i18n';

const t = i18n.getFixedT('en', 'common');

describe('AnnualTargetHours', () => {
  it('should render a table with correct headers', () => {
    const { getByRole, baseElement } = render(
      <AnnualTargetHours targetData={null} accrualsData={null} />
    );

    const table = getByRole('table');
    expect(table).toBeTruthy();

    expect(baseElement.innerHTML).toMatchSnapshot();
  });
});

it('should render a table with en-gb headers', () => {
  const { getByRole } = render(
    <AnnualTargetHours targetData={null} accrualsData={null} />
  );

  const table = getByRole('table');
  expect(table).toBeTruthy();

  const headers = ['Total', 'Worked', 'Remaining', 'Target'];

  headers.map((header) => {
    expect(screen.getByText(header)).toBeTruthy();
  });
});
