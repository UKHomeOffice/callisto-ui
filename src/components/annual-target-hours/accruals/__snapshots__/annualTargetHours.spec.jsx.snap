// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnnualTargetHours should render a table with correct headers 1`] = `"<div><div class=\\"accruals-container\\"><form class=\\"grey-border\\"><h1 class=\\"govuk-heading-m newline\\">No agreement has been found</h1><table class=\\"govuk-table\\"><tbody class=\\"govuk-table__body\\"><tr class=\\"govuk-table__row\\"><th scope=\\"row\\" class=\\"govuk-table__header\\">Total</th><td class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td></tr><tr class=\\"govuk-table__row\\"><th scope=\\"row\\" class=\\"govuk-table__header\\">Worked</th><td class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td></tr><tr class=\\"govuk-table__row\\"><th scope=\\"row\\" class=\\"govuk-table__header\\">Remaining</th><td class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td></tr><tr class=\\"govuk-table__row\\"><th scope=\\"row\\" class=\\"govuk-table__header\\">Target</th><td class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td></tr></tbody></table></form></div></div>"`;
